### PR TITLE
Replace assertion with exception

### PR DESCRIPTION
**Pull request by anmolsjoshi**
# What does this PR do?

Replaces assert with Exceptions as per https://github.com/huggingface/transformers/issues/12789.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger

### CODE REVIEW

File: src/transformers/models/ctrl/modeling_tf_ctrl.py [hunk 1]
```diff
@@ -816,9 +816,8 @@ def call(
                 batch_size, sequence_length = shape_list(input_ids)[:2]
             else:
                 batch_size, sequence_length = shape_list(inputs_embeds)[:2]
-            assert (
-                self.config.pad_token_id is not None or batch_size == 1
-            ), "Cannot handle batch sizes > 1 if no padding token is defined."
+            if self.config.pad_token_id is None or batch_size != 1:
```
Comment (sgugger) [inline src/transformers/models/ctrl/modeling_tf_ctrl.py:819]:
```suggestion
            if self.config.pad_token_id is not None and batch_size != 1:
```
Careful, the negation is wrong here :-)

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 2]
```diff
@@ -875,9 +872,8 @@ def forward(
         else:
             batch_size, sequence_length = inputs_embeds.shape[:2]
 
-        assert (
-            self.config.pad_token_id is not None or batch_size == 1
-        ), "Cannot handle batch sizes > 1 if no padding token is defined."
+        if self.config.pad_token_id is None or batch_size != 1:
```
Comment (sgugger) [inline src/transformers/models/gpt_neo/modeling_gpt_neo.py:875]:
```suggestion
        if self.config.pad_token_id is None and batch_size != 1:
```

File: src/transformers/models/ctrl/modeling_ctrl.py [hunk 3]
```diff
@@ -669,9 +670,8 @@ def forward(
         else:
             batch_size, sequence_length = inputs_embeds.shape[:2]
 
-        assert (
-            self.config.pad_token_id is not None or batch_size == 1
-        ), "Cannot handle batch sizes > 1 if no padding token is defined."
+        if self.config.pad_token_id is None or batch_size != 1:
```
Comment (sgugger) [inline src/transformers/models/ctrl/modeling_ctrl.py:673]:
```suggestion
        if self.config.pad_token_id is None and batch_size != 1:
```

File: src/transformers/models/gptj/modeling_gptj.py [hunk 4]
```diff
@@ -939,9 +940,8 @@ def forward(
         else:
             batch_size = inputs_embeds.shape[0]
 
-        assert (
-            self.config.pad_token_id is not None or batch_size == 1
-        ), "Cannot handle batch sizes > 1 if no padding token is defined."
+        if self.config.pad_token_id is None or batch_size != 1:
```
Comment (sgugger) [inline src/transformers/models/gptj/modeling_gptj.py:943]:
```suggestion
        if self.config.pad_token_id is None and batch_size != 1:
```

File: src/transformers/models/gptj/modeling_tf_gptj.py [hunk 5]
```diff
@@ -934,9 +934,8 @@ def call(
         loss = None
 
         if labels is not None:
-            assert (
-                self.config.pad_token_id is not None or logits_shape[0] == 1
-            ), "Cannot handle batch sizes > 1 if no padding token is defined."
+            if self.config.pad_token_id is None or logits_shape[0] != 1:
```
Comment (sgugger) [inline src/transformers/models/gptj/modeling_tf_gptj.py:937]:
```suggestion
            if self.config.pad_token_id is None and logits_shape[0] != 1:
```

File: src/transformers/models/led/modeling_led.py [hunk 6]
```diff
@@ -833,27 +835,34 @@ def forward(
         src_len = key_states.size(1)
         attn_weights = torch.bmm(query_states, key_states.transpose(1, 2))
 
-        assert attn_weights.size() == (
+        if attn_weights.size() != (
             bsz * self.num_heads,
             tgt_len,
             src_len,
-        ), f"Attention weights should be of size {(bsz * self.num_heads, tgt_len, src_len)}, but is {attn_weights.size()}"
+        ):
```
Comment (sgugger) [inline src/transformers/models/led/modeling_led.py:842]: This can all fit in one line now.

File: src/transformers/models/led/modeling_led.py [hunk 7]
```diff
@@ -833,27 +835,34 @@ def forward(
...
 
         if attention_mask is not None:
-            assert attention_mask.size() == (
+            if attention_mask.size() != (
                 bsz,
                 1,
                 tgt_len,
                 src_len,
-            ), f"Attention mask should be of size {(bsz, 1, tgt_len, src_len)}, but is {attention_mask.size()}"
+            ):
```
Comment (sgugger) [inline src/transformers/models/led/modeling_led.py:853]: Same here.

File: src/transformers/models/led/modeling_led.py [hunk 8]
```diff
@@ -871,11 +880,14 @@ def forward(
 
         attn_output = torch.bmm(attn_probs, value_states)
 
-        assert attn_output.size() == (
+        if attn_output.size() != (
             bsz * self.num_heads,
             tgt_len,
             self.head_dim,
-        ), f"`attn_output` should be of size {(bsz, self.num_heads, tgt_len, self.head_dim)}, but is {attn_output.size()}"
+        ):
```
Comment (sgugger) [inline src/transformers/models/led/modeling_led.py:887]: And there.

**Review by sgugger — APPROVED** [review]
Wow, that's a lot of work, thanks a lot! Could you a quick search for the pattern `if self.config.pad_token_id is None or logits_shape[0] != 1:` in case I missed any? The `or` needs to be replaced by an `and`.

Also, since it touches so many modeling files, I'd like a second review on this @LysandreJik if you can.

**Review by LysandreJik — APPROVED** [review]
Impressive PR! Thanks for working on it, @anmolsjoshi. Apart from Sylvain's comments, didn't find anything else that needs to be changed.

**Comment by anmolsjoshi** [comment]
@sgugger @LysandreJik thanks for the review! I have incorporated all the requested changes.

**Comment by anmolsjoshi** [comment]
@sgugger unsure why the PR Documentation check is failing

**Comment by sgugger** [comment]
Failure is unrelated to this PR and is fixed independently.
Thanks a lot for addressing all comments!

**Pull request merged**